src/PowerShellEditorServices/Language/RunspaceSynchronizer.cs

```diff
@@ -0,0 +1,272 @@
+            targetEngineIntrinsics = ReflectionUtils.GetEngineIntrinsics(runspaceTarget);
+            IsReadyForEvents = true;
+
+            sourceEngineIntrinsics.Events.SubscribeEvent(
+                source: null,
+                eventName: null,
+                sourceIdentifier: PSEngineEvent.OnIdle.ToString(),
+                data: null,
+                handlerDelegate: HandleRunspaceStateChange,
+                supportEvent: true,
+                forwardEvent: false);
+
+            Activate();
+            // Trigger events
+            HandleRunspaceStateChange(sender: null, args: null);
+        }
+
+        /// <summary>
+        /// Does a thing
+        /// </summary>
+        public static void Activate()
+        {
+            SourceActionEnabled = true;
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private static void HandleRunspaceStateChange(object sender, PSEventArgs args)
+        {
+            if (!SourceActionEnabled)
+            {
+                return;
+            }
+
+            SourceActionEnabled = false;
+
+            var newOrChangedModules = new List<PSModuleInfo>();
+            List<PSModuleInfo> modules = ReflectionUtils.GetModules(sourceRunspace);
+            foreach (PSModuleInfo module in modules)
+            {
+                if (moduleCache.Add(module))
+                {
+                    newOrChangedModules.Add(module);
+                }
+            }
+
+
+            var newOrChangedVars = new List<PSVariable>();
+
+            var variables = sourceEngineIntrinsics.GetVariables();
+            foreach (var variable in variables)
+            {
+                // If the variable is a magic variable or it's type has not changed, then skip it.
+                if(POWERSHELL_MAGIC_VARIABLES.Contains(variable.Name) ||
+                    (variableCache.TryGetValue(variable.Name, out Type value) && value == variable.Value?.GetType()))
+                {
+                    continue;
+                }
+
+                // Add the variable to the cache and mark it as a newOrChanged variable.
+                variableCache[variable.Name] = variable.Value?.GetType();
+                newOrChangedVars.Add(variable);
+            }
+
+            if (newOrChangedModules.Count > 0)
+            {
+                // Import the modules in the targetRunspace with -Force
+                using (PowerShell pwsh = PowerShell.Create())
+                {
+                    pwsh.Runspace = targetRunspace;
+
+                    foreach (PSModuleInfo moduleInfo in newOrChangedModules)
+                    {
+                        if(moduleInfo.Path != null)
+                        {
+                            pwsh.AddCommand("Import-Module")
+                                .AddParameter("Name", moduleInfo.Path)
+                                .AddParameter("Force")
+                                .AddStatement();
```

Comment:
This won't work for argument completers that depend on module state.  Also not related to this line but argument completers not loaded by a module won't be transferred either.

Reply:
This will cause some modules to throw.  iirc a commonly installed version of the VMWare module does not handle being imported into multiple runspaces well.  There's a few modules like that.

Reply:
> Also not related to this line but argument completers not loaded by a module won't be transferred either.

Can you elaborate the scenario for this one? The assembly that implements a argument completer not getting loaded by the module?

Reply:
> This will cause some modules to throw. iirc a commonly installed version of the VMWare module does not handle being imported into multiple runspaces well. There's a few modules like that.

This is a valid concern. Normally a module shouldn't be made only work when importing once in a process ... but we need to evaluate how bad it would be (a black list?)

Reply:
> Also not related to this line but argument completers not loaded by a module won't be transferred either

@rjmholt thought of the idea of overriding `Register-ArgumentCompleter` which will forward the call to the target runspace. That said, perhaps we should also ask ourselves, could running argument completers for Intellisense be a potential security concern and a valid thing to skip in the _editor_ intellisense at least.

Reply:
> Can you elaborate the scenario for this one? The assembly that implements a argument completer not getting loaded by the module?

Sorry that was confusing in the context of this line, I mean more generally.  Like someone just dot sourcing a script or manually running `Register-ArgumentCompleter`.

Reply:
> That said, perhaps we should also ask ourselves, could running argument completers for Intellisense be a potential security concern and a valid thing to skip in the _editor_ intellisense at least.

Considering they have to run code in order to register an argument completer I don't think it's a security issue.  We do automatically import modules during intellisense which could register an argument completer.  *That* might be a security issue, but more because loading a module runs code, not necessarily related to the fact that a completer could be loaded.

Reply:
I may override this cmdlet. It's not as complicated and high profile as `Import-Module`